# E-COMMERCE PURCHASE INTENTION MODEL

#### This models designed to analyse web analystic data and predict whether  a customer will buy product or not during their visit

## Importing Main Libraries

In [40]:
import time
import pandas as pd
import numpy as np
# import matplotlib.pyplot as plt
# import seaborn as sns  

from sklearn.preprocessing import OrdinalEncoder
from sklearn.model_selection import train_test_split

from sklearn.impute import SimpleImputer
from sklearn.preprocessing import OneHotEncoder
from sklearn.compose import ColumnTransformer
from imblearn.over_sampling import SMOTE
from sklearn.preprocessing import MinMaxScaler

from sklearn.feature_selection import SelectKBest
from sklearn.feature_selection import chi2

from imblearn.pipeline import Pipeline as imbpipeline
from sklearn.pipeline import Pipeline

from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.linear_model import RidgeClassifier
from sklearn.naive_bayes import BernoulliNB
from sklearn.svm import SVC
from xgboost import XGBClassifier
# from sklearn.dummy import DummyClassifier
# from sklearn.linear_model import SGDClassifier
# from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import cross_val_score

from sklearn.metrics import accuracy_score
from sklearn.metrics import roc_auc_score
from sklearn.metrics import f1_score
from sklearn.metrics import classification_report
# from sklearn.metrics import confusion_matrix


# - - - - - - -  To ignore warning   - - - - - - - #

from warnings import simplefilter
from sklearn.exceptions import ConvergenceWarning
simplefilter("ignore", category=ConvergenceWarning)

# Loading the Dataset

In [2]:
df=pd.read_csv("online_shoppers_intention.csv")

In [3]:
df

Administrative  Administrative_Duration  Informational  \
0                   0                      0.0              0   
1                   0                      0.0              0   
2                   0                      0.0              0   
3                   0                      0.0              0   
4                   0                      0.0              0   
...               ...                      ...            ...   
12325               3                    145.0              0   
12326               0                      0.0              0   
12327               0                      0.0              0   
12328               4                     75.0              0   
12329               0                      0.0              0   

       Informational_Duration  ProductRelated  ProductRelated_Duration  \
0                         0.0               1                 0.000000   
1                         0.0               2                64.000000   
2                         0.0               1                 0.000000   
3                         0.0               2                 2.666667   
4                         0.0              10               627.500000   
...                       ...             ...                      ...   
12325                     0.0              53              1783.791667   
12326                     0.0               5               465.750000   
12327                     0.0               6               184.250000   
12328                     0.0              15               346.000000   
12329                     0.0               3                21.250000   

       BounceRates  ExitRates  PageValues  SpecialDay Month  OperatingSystems  \
0         0.200000   0.200000    0.000000         0.0   Feb                 1   
1         0.000000   0.100000    0.000000         0.0   Feb                 2   
2         0.200000   0.200000    0.000000         0.0   Feb                 4   
3         0.050000   0.140000    0.000000         0.0   Feb                 3   
4         0.020000   0.050000    0.000000         0.0   Feb                 3   
...            ...        ...         ...         ...   ...               ...   
12325     0.007143   0.029031   12.241717         0.0   Dec                 4   
12326     0.000000   0.021333    0.000000         0.0   Nov                 3   
12327     0.083333   0.086667    0.000000         0.0   Nov                 3   
12328     0.000000   0.021053    0.000000         0.0   Nov                 2   
12329     0.000000   0.066667    0.000000         0.0   Nov                 3   

       Browser  Region  TrafficType        VisitorType  Weekend  Revenue  
0            1       1            1  Returning_Visitor    False    False  
1            2       1            2  Returning_Visitor    False    False  
2            1       9            3  Returning_Visitor    False    False  
3            2       2            4  Returning_Visitor    False    False  
4            3       1            4  Returning_Visitor     True    False  
...        ...     ...          ...                ...      ...      ...  
12325        6       1            1  Returning_Visitor     True    False  
12326        2       1            8  Returning_Visitor     True    False  
12327        2       1           13  Returning_Visitor     True    False  
12328        2       3           11  Returning_Visitor    False    False  
12329        2       1            2        New_Visitor     True    False  

[12330 rows x 18 columns]

In [4]:
# Showing first five rows of the DataFrame
df.head()

Administrative  Administrative_Duration  Informational  \
0               0                      0.0              0   
1               0                      0.0              0   
2               0                      0.0              0   
3               0                      0.0              0   
4               0                      0.0              0   

   Informational_Duration  ProductRelated  ProductRelated_Duration  \
0                     0.0               1                 0.000000   
1                     0.0               2                64.000000   
2                     0.0               1                 0.000000   
3                     0.0               2                 2.666667   
4                     0.0              10               627.500000   

   BounceRates  ExitRates  PageValues  SpecialDay Month  OperatingSystems  \
0         0.20       0.20         0.0         0.0   Feb                 1   
1         0.00       0.10         0.0         0.0   Feb                 2   
2         0.20       0.20         0.0         0.0   Feb                 4   
3         0.05       0.14         0.0         0.0   Feb                 3   
4         0.02       0.05         0.0         0.0   Feb                 3   

   Browser  Region  TrafficType        VisitorType  Weekend  Revenue  
0        1       1            1  Returning_Visitor    False    False  
1        2       1            2  Returning_Visitor    False    False  
2        1       9            3  Returning_Visitor    False    False  
3        2       2            4  Returning_Visitor    False    False  
4        3       1            4  Returning_Visitor     True    False

In [5]:
# Showing last five rows of the DataFrame
df.tail()

Administrative  Administrative_Duration  Informational  \
12325               3                    145.0              0   
12326               0                      0.0              0   
12327               0                      0.0              0   
12328               4                     75.0              0   
12329               0                      0.0              0   

       Informational_Duration  ProductRelated  ProductRelated_Duration  \
12325                     0.0              53              1783.791667   
12326                     0.0               5               465.750000   
12327                     0.0               6               184.250000   
12328                     0.0              15               346.000000   
12329                     0.0               3                21.250000   

       BounceRates  ExitRates  PageValues  SpecialDay Month  OperatingSystems  \
12325     0.007143   0.029031   12.241717         0.0   Dec                 4   
12326     0.000000   0.021333    0.000000         0.0   Nov                 3   
12327     0.083333   0.086667    0.000000         0.0   Nov                 3   
12328     0.000000   0.021053    0.000000         0.0   Nov                 2   
12329     0.000000   0.066667    0.000000         0.0   Nov                 3   

       Browser  Region  TrafficType        VisitorType  Weekend  Revenue  
12325        6       1            1  Returning_Visitor     True    False  
12326        2       1            8  Returning_Visitor     True    False  
12327        2       1           13  Returning_Visitor     True    False  
12328        2       3           11  Returning_Visitor    False    False  
12329        2       1            2        New_Visitor     True    False

In [6]:
# Showing no of rows in  dataframe
len(df)

12330

## Clean data - Feature Engineering

In [7]:
# First we can check our data information by using info() method
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12330 entries, 0 to 12329
Data columns (total 18 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Administrative           12330 non-null  int64  
 1   Administrative_Duration  12330 non-null  float64
 2   Informational            12330 non-null  int64  
 3   Informational_Duration   12330 non-null  float64
 4   ProductRelated           12330 non-null  int64  
 5   ProductRelated_Duration  12330 non-null  float64
 6   BounceRates              12330 non-null  float64
 7   ExitRates                12330 non-null  float64
 8   PageValues               12330 non-null  float64
 9   SpecialDay               12330 non-null  float64
 10  Month                    12330 non-null  object 
 11  OperatingSystems         12330 non-null  int64  
 12  Browser                  12330 non-null  int64  
 13  Region                   12330 non-null  int64  
 14  TrafficType           

In [8]:
#The Weekend and Revenue columns are currently set to Boolean values,so we first need to convert into binary values.
df["Weekend"].unique()

array([False,  True])

In [9]:
df["Revenue"].unique()

array([False,  True])

In [10]:
df["Weekend"]=df["Weekend"].replace((False,True),(0,1))

In [11]:
df.head()

Administrative  Administrative_Duration  Informational  \
0               0                      0.0              0   
1               0                      0.0              0   
2               0                      0.0              0   
3               0                      0.0              0   
4               0                      0.0              0   

   Informational_Duration  ProductRelated  ProductRelated_Duration  \
0                     0.0               1                 0.000000   
1                     0.0               2                64.000000   
2                     0.0               1                 0.000000   
3                     0.0               2                 2.666667   
4                     0.0              10               627.500000   

   BounceRates  ExitRates  PageValues  SpecialDay Month  OperatingSystems  \
0         0.20       0.20         0.0         0.0   Feb                 1   
1         0.00       0.10         0.0         0.0   Feb                 2   
2         0.20       0.20         0.0         0.0   Feb                 4   
3         0.05       0.14         0.0         0.0   Feb                 3   
4         0.02       0.05         0.0         0.0   Feb                 3   

   Browser  Region  TrafficType        VisitorType  Weekend  Revenue  
0        1       1            1  Returning_Visitor        0    False  
1        2       1            2  Returning_Visitor        0    False  
2        1       9            3  Returning_Visitor        0    False  
3        2       2            4  Returning_Visitor        0    False  
4        3       1            4  Returning_Visitor        1    False

In [12]:
df["Revenue"]=df["Revenue"].replace((False,True),(0,1))

In [13]:
df.head()

Administrative  Administrative_Duration  Informational  \
0               0                      0.0              0   
1               0                      0.0              0   
2               0                      0.0              0   
3               0                      0.0              0   
4               0                      0.0              0   

   Informational_Duration  ProductRelated  ProductRelated_Duration  \
0                     0.0               1                 0.000000   
1                     0.0               2                64.000000   
2                     0.0               1                 0.000000   
3                     0.0               2                 2.666667   
4                     0.0              10               627.500000   

   BounceRates  ExitRates  PageValues  SpecialDay Month  OperatingSystems  \
0         0.20       0.20         0.0         0.0   Feb                 1   
1         0.00       0.10         0.0         0.0   Feb                 2   
2         0.20       0.20         0.0         0.0   Feb                 4   
3         0.05       0.14         0.0         0.0   Feb                 3   
4         0.02       0.05         0.0         0.0   Feb                 3   

   Browser  Region  TrafficType        VisitorType  Weekend  Revenue  
0        1       1            1  Returning_Visitor        0        0  
1        2       1            2  Returning_Visitor        0        0  
2        1       9            3  Returning_Visitor        0        0  
3        2       2            4  Returning_Visitor        0        0  
4        3       1            4  Returning_Visitor        1        0

In [14]:
# Let’s understand VisitorType column
df["VisitorType"].value_counts()

VisitorType
Returning_Visitor    10551
New_Visitor           1694
Other                   85
Name: count, dtype: int64

In [15]:
# VisitorType contains either Returning_Visitor or New_Visitor.So Let’s add Returning_Visitor column to existing dataframe

condition=df["VisitorType"]=="Returning_Visitor"
df["Returning_Visitor"]=np.where(condition,1,0)


In [16]:
df.head()

Administrative  Administrative_Duration  Informational  \
0               0                      0.0              0   
1               0                      0.0              0   
2               0                      0.0              0   
3               0                      0.0              0   
4               0                      0.0              0   

   Informational_Duration  ProductRelated  ProductRelated_Duration  \
0                     0.0               1                 0.000000   
1                     0.0               2                64.000000   
2                     0.0               1                 0.000000   
3                     0.0               2                 2.666667   
4                     0.0              10               627.500000   

   BounceRates  ExitRates  PageValues  SpecialDay Month  OperatingSystems  \
0         0.20       0.20         0.0         0.0   Feb                 1   
1         0.00       0.10         0.0         0.0   Feb                 2   
2         0.20       0.20         0.0         0.0   Feb                 4   
3         0.05       0.14         0.0         0.0   Feb                 3   
4         0.02       0.05         0.0         0.0   Feb                 3   

   Browser  Region  TrafficType        VisitorType  Weekend  Revenue  \
0        1       1            1  Returning_Visitor        0        0   
1        2       1            2  Returning_Visitor        0        0   
2        1       9            3  Returning_Visitor        0        0   
3        2       2            4  Returning_Visitor        0        0   
4        3       1            4  Returning_Visitor        1        0   

   Returning_Visitor  
0                  1  
1                  1  
2                  1  
3                  1  
4                  1

In [17]:
# Remove VisitorType columns
df.drop("VisitorType",axis=1,inplace=True)

In [18]:
df.head()

Administrative  Administrative_Duration  Informational  \
0               0                      0.0              0   
1               0                      0.0              0   
2               0                      0.0              0   
3               0                      0.0              0   
4               0                      0.0              0   

   Informational_Duration  ProductRelated  ProductRelated_Duration  \
0                     0.0               1                 0.000000   
1                     0.0               2                64.000000   
2                     0.0               1                 0.000000   
3                     0.0               2                 2.666667   
4                     0.0              10               627.500000   

   BounceRates  ExitRates  PageValues  SpecialDay Month  OperatingSystems  \
0         0.20       0.20         0.0         0.0   Feb                 1   
1         0.00       0.10         0.0         0.0   Feb                 2   
2         0.20       0.20         0.0         0.0   Feb                 4   
3         0.05       0.14         0.0         0.0   Feb                 3   
4         0.02       0.05         0.0         0.0   Feb                 3   

   Browser  Region  TrafficType  Weekend  Revenue  Returning_Visitor  
0        1       1            1        0        0                  1  
1        2       1            2        0        0                  1  
2        1       9            3        0        0                  1  
3        2       2            4        0        0                  1  
4        3       1            4        1        0                  1

In [19]:
# Checking all columns data type
df.dtypes

Administrative               int64
Administrative_Duration    float64
Informational                int64
Informational_Duration     float64
ProductRelated               int64
ProductRelated_Duration    float64
BounceRates                float64
ExitRates                  float64
PageValues                 float64
SpecialDay                 float64
Month                       object
OperatingSystems             int64
Browser                      int64
Region                       int64
TrafficType                  int64
Weekend                      int64
Revenue                      int64
Returning_Visitor            int32
dtype: object

In [20]:
# Month columns have object data type means string type,Lest apply Ordinal Encoding over this column.

df["Month"].unique()

array(['Feb', 'Mar', 'May', 'Oct', 'June', 'Jul', 'Aug', 'Nov', 'Sep',
       'Dec'], dtype=object)

In [21]:
ordinal_encoder=OrdinalEncoder()
df["Month"]=ordinal_encoder.fit_transform(df[["Month"]])

In [22]:
df["Month"].unique()

array([2., 5., 6., 8., 4., 3., 0., 7., 9., 1.])

In [23]:
df.dtypes

Administrative               int64
Administrative_Duration    float64
Informational                int64
Informational_Duration     float64
ProductRelated               int64
ProductRelated_Duration    float64
BounceRates                float64
ExitRates                  float64
PageValues                 float64
SpecialDay                 float64
Month                      float64
OperatingSystems             int64
Browser                      int64
Region                       int64
TrafficType                  int64
Weekend                      int64
Revenue                      int64
Returning_Visitor            int32
dtype: object

In [24]:
# Revenue column is the target variable,Let’s check the value_counts() on this column.
df["Revenue"].value_counts()

Revenue
0    10422
1     1908
Name: count, dtype: int64

In [25]:
# Access required columns
result=df.columns[1:]
result

Index(['Administrative_Duration', 'Informational', 'Informational_Duration',
       'ProductRelated', 'ProductRelated_Duration', 'BounceRates', 'ExitRates',
       'PageValues', 'SpecialDay', 'Month', 'OperatingSystems', 'Browser',
       'Region', 'TrafficType', 'Weekend', 'Revenue', 'Returning_Visitor'],
      dtype='object')

In [26]:
result=df[df.columns[1:]]
result

Administrative_Duration  Informational  Informational_Duration  \
0                          0.0              0                     0.0   
1                          0.0              0                     0.0   
2                          0.0              0                     0.0   
3                          0.0              0                     0.0   
4                          0.0              0                     0.0   
...                        ...            ...                     ...   
12325                    145.0              0                     0.0   
12326                      0.0              0                     0.0   
12327                      0.0              0                     0.0   
12328                     75.0              0                     0.0   
12329                      0.0              0                     0.0   

       ProductRelated  ProductRelated_Duration  BounceRates  ExitRates  \
0                   1                 0.000000     0.200000   0.200000   
1                   2                64.000000     0.000000   0.100000   
2                   1                 0.000000     0.200000   0.200000   
3                   2                 2.666667     0.050000   0.140000   
4                  10               627.500000     0.020000   0.050000   
...               ...                      ...          ...        ...   
12325              53              1783.791667     0.007143   0.029031   
12326               5               465.750000     0.000000   0.021333   
12327               6               184.250000     0.083333   0.086667   
12328              15               346.000000     0.000000   0.021053   
12329               3                21.250000     0.000000   0.066667   

       PageValues  SpecialDay  Month  OperatingSystems  Browser  Region  \
0        0.000000         0.0    2.0                 1        1       1   
1        0.000000         0.0    2.0                 2        2       1   
2        0.000000         0.0    2.0                 4        1       9   
3        0.000000         0.0    2.0                 3        2       2   
4        0.000000         0.0    2.0                 3        3       1   
...           ...         ...    ...               ...      ...     ...   
12325   12.241717         0.0    1.0                 4        6       1   
12326    0.000000         0.0    7.0                 3        2       1   
12327    0.000000         0.0    7.0                 3        2       1   
12328    0.000000         0.0    7.0                 2        2       3   
12329    0.000000         0.0    7.0                 3        2       1   

       TrafficType  Weekend  Revenue  Returning_Visitor  
0                1        0        0                  1  
1                2        0        0                  1  
2                3        0        0                  1  
3                4        0        0                  1  
4                4        1        0                  1  
...            ...      ...      ...                ...  
12325            1        1        0                  1  
12326            8        1        0                  1  
12327           13        1        0                  1  
12328           11        0        0                  1  
12329            2        1        0                  0  

[12330 rows x 17 columns]

In [27]:
result=df[df.columns[1:]].corr()
# dataframe.corr() is used to find the pairwise correlation of all columns in the dataframe. 
# Any na values are automatically excluded. For any non-numeric data type columns in the dataframe 
# it is ignored.

In [28]:
result

Administrative_Duration  Informational  \
Administrative_Duration                 1.000000       0.302710   
Informational                           0.302710       1.000000   
Informational_Duration                  0.238031       0.618955   
ProductRelated                          0.289087       0.374164   
ProductRelated_Duration                 0.355422       0.387505   
BounceRates                            -0.144170      -0.116114   
ExitRates                              -0.205798      -0.163666   
PageValues                              0.067608       0.048632   
SpecialDay                             -0.073304      -0.048219   
Month                                   0.029061       0.019743   
OperatingSystems                       -0.007343      -0.009527   
Browser                                -0.015392      -0.038235   
Region                                 -0.005561      -0.029169   
TrafficType                            -0.014376      -0.034491   
Weekend                                 0.014990       0.035785   
Revenue                                 0.093587       0.095200   
Returning_Visitor                      -0.022525       0.057399   

                         Informational_Duration  ProductRelated  \
Administrative_Duration                0.238031        0.289087   
Informational                          0.618955        0.374164   
Informational_Duration                 1.000000        0.280046   
ProductRelated                         0.280046        1.000000   
ProductRelated_Duration                0.347364        0.860927   
BounceRates                           -0.074067       -0.204578   
ExitRates                             -0.105276       -0.292526   
PageValues                             0.030861        0.056282   
SpecialDay                            -0.030577       -0.023958   
Month                                  0.005987        0.070299   
OperatingSystems                      -0.009579        0.004290   
Browser                               -0.019285       -0.013146   
Region                                -0.027144       -0.038122   
TrafficType                           -0.024675       -0.043064   
Weekend                                0.024078        0.016092   
Revenue                                0.070345        0.158538   
Returning_Visitor                      0.045501        0.128738   

                         ProductRelated_Duration  BounceRates  ExitRates  \
Administrative_Duration                 0.355422    -0.144170  -0.205798   
Informational                           0.387505    -0.116114  -0.163666   
Informational_Duration                  0.347364    -0.074067  -0.105276   
ProductRelated                          0.860927    -0.204578  -0.292526   
ProductRelated_Duration                 1.000000    -0.184541  -0.251984   
BounceRates                            -0.184541     1.000000   0.913004   
ExitRates                              -0.251984     0.913004   1.000000   
PageValues                              0.052823    -0.119386  -0.174498   
SpecialDay                             -0.036380     0.072702   0.102242   
Month                                   0.061186    -0.023763  -0.039049   
OperatingSystems                        0.002976     0.023823   0.014567   
Browser                                -0.007380    -0.015772  -0.004442   
Region                                 -0.033091    -0.006485  -0.008907   
TrafficType                            -0.036377     0.078286   0.078616   
Weekend                                 0.007311    -0.046514  -0.062587   
Revenue                                 0.152373    -0.150673  -0.207071   
Returning_Visitor                       0.120489     0.129908   0.171987   

                         PageValues  SpecialDay     Month  OperatingSystems  \
Administrative_Duration    0.067608   -0.073304  0.029061         -0.007343   
Informational              0.048632   -0.048219  0.019743         -0.009527   
Infor

In [29]:
result=df[df.columns[1:]].corr()["Revenue"]

In [30]:
result

Administrative_Duration    0.093587
Informational              0.095200
Informational_Duration     0.070345
ProductRelated             0.158538
ProductRelated_Duration    0.152373
BounceRates               -0.150673
ExitRates                 -0.207071
PageValues                 0.492569
SpecialDay                -0.082305
Month                      0.080150
OperatingSystems          -0.014668
Browser                    0.023984
Region                    -0.011595
TrafficType               -0.005113
Weekend                    0.029295
Revenue                    1.000000
Returning_Visitor         -0.103843
Name: Revenue, dtype: float64

In [31]:
result1=result.sort_values(ascending=False)

In [32]:
result1

Revenue                    1.000000
PageValues                 0.492569
ProductRelated             0.158538
ProductRelated_Duration    0.152373
Informational              0.095200
Administrative_Duration    0.093587
Month                      0.080150
Informational_Duration     0.070345
Weekend                    0.029295
Browser                    0.023984
TrafficType               -0.005113
Region                    -0.011595
OperatingSystems          -0.014668
SpecialDay                -0.082305
Returning_Visitor         -0.103843
BounceRates               -0.150673
ExitRates                 -0.207071
Name: Revenue, dtype: float64

## Create the training and test data

*  Now we need to prepare the feature set, we need to define X and y.
* The X feature set will include all the features we created above
* The y feature having target variable alone.

In [33]:
X=df.drop("Revenue",axis=1)
y=df.Revenue

In [34]:
X

Administrative  Administrative_Duration  Informational  \
0                   0                      0.0              0   
1                   0                      0.0              0   
2                   0                      0.0              0   
3                   0                      0.0              0   
4                   0                      0.0              0   
...               ...                      ...            ...   
12325               3                    145.0              0   
12326               0                      0.0              0   
12327               0                      0.0              0   
12328               4                     75.0              0   
12329               0                      0.0              0   

       Informational_Duration  ProductRelated  ProductRelated_Duration  \
0                         0.0               1                 0.000000   
1                         0.0               2                64.000000   
2                         0.0               1                 0.000000   
3                         0.0               2                 2.666667   
4                         0.0              10               627.500000   
...                       ...             ...                      ...   
12325                     0.0              53              1783.791667   
12326                     0.0               5               465.750000   
12327                     0.0               6               184.250000   
12328                     0.0              15               346.000000   
12329                     0.0               3                21.250000   

       BounceRates  ExitRates  PageValues  SpecialDay  Month  \
0         0.200000   0.200000    0.000000         0.0    2.0   
1         0.000000   0.100000    0.000000         0.0    2.0   
2         0.200000   0.200000    0.000000         0.0    2.0   
3         0.050000   0.140000    0.000000         0.0    2.0   
4         0.020000   0.050000    0.000000         0.0    2.0   
...            ...        ...         ...         ...    ...   
12325     0.007143   0.029031   12.241717         0.0    1.0   
12326     0.000000   0.021333    0.000000         0.0    7.0   
12327     0.083333   0.086667    0.000000         0.0    7.0   
12328     0.000000   0.021053    0.000000         0.0    7.0   
12329     0.000000   0.066667    0.000000         0.0    7.0   

       OperatingSystems  Browser  Region  TrafficType  Weekend  \
0                     1        1       1            1        0   
1                     2        2       1            2        0   
2                     4        1       9            3        0   
3                     3        2       2            4        0   
4                     3        3       1            4        1   
...                 ...      ...     ...          ...      ...   
12325                 4        6       1            1        1   
12326                 3        2       1            8        1   
12327                 3        2       1           13        1   
12328                 2        2       3           11        0   
12329                 3        2       1            2        1   

       Returning_Visitor  
0                      1  
1                      1  
2                      1  
3                      1  
4                      1  
...                  ...  
12325                  1  
12326                  1  
12327                  1  
12328                  1  
12329                  0  

[12330 rows x 17 columns]

In [35]:
y

0        0
1        0
2        0
3        0
4        0
        ..
12325    0
12326    0
12327    0
12328    0
12329    0
Name: Revenue, Length: 12330, dtype: int64

## Train and Test data

- We need to split(randomly) data into training (70%) and testing (30%) datasets.
- We can split data by using train_test_split(X, y, test_size = 0.3, random_state) function.
- The random_state value ensures we get reproducible results each time we run the code.

In [36]:
X_train,X_test,y_train,y_test=train_test_split(X,y,train_size=0.33,random_state=0)

## Model Pipeline

## A Machine Learning pipeline
* This is the process of automating the workflow of a complete machine learning task
* This is having group of steps like data to be transformed and correlated together in a model that can be analysed to get the output
* A typical pipeline includes raw data input, features, outputs, model parameters, ML models, and Predictions
* Pipeline also having sequential steps that perform everything like data extraction and pre-processing to model training and deployment in Machine learning in a modular approach
* It means that in the pipeline, each step is designed as an independent module, and all these modules are tied together to get the final result

## Create a model pipeline

 * Now we need to create a model pipeline for our project
 * This Pipeline handles the encoding of data using the ColumnTransformer()
 * By using this we can even avoid like manually generating some of the features we created above
 * This also imputes any missing values with realistic values
 * It scales the data before we pass it to the model

## SelectKBest and SMOTE
* By using SelectKBest() we can select the optimal features. From this we are getting around 6 features as per shown in Pearson correlation
* By using SMOTE(Synthetic Minority Oversampling Technique) we can handle class imbalance

## model_pipeline(X, model) function explanation
* model_pipleline(X, model) is a user defined function for this project
* This function returns pipeline to pre-process data and bundle with a model
* There are two arguments for this function X means X_train data and model means model object Ex: XGBClassifier object

In [37]:
def model_pipeline(X,model):
    n_c=X.select_dtypes(exclude=["object"]).columns.values.tolist()
    c_c=X.select_dtypes(include=["object"]).columns.values.tolist()

    numeric_columns = n_c
    categorical_columns=c_c

    numeric_pipeline = SimpleImputer(strategy='constant')
    categorical_pipeline = OneHotEncoder(handle_unknown='ignore')
        
    a = ('numeric', numeric_pipeline, numeric_columns)
    
    b = ('categorical', categorical_pipeline, categorical_columns)

    preprocessor = ColumnTransformer(transformers=[a, b], remainder='passthrough')
    
    c = ('preprocessor', preprocessor)
    
    d = ('smote', SMOTE(random_state=1))
    
    e = ('scaler', MinMaxScaler())
    
    f = ('feature_selection', SelectKBest(score_func=chi2, k=6))
    
    g = ('model', model)

    bundled_pipeline = imbpipeline(steps=[c, d, e, f, g])

    return bundled_pipeline

## Select best model

* Generally we can select single model to proceeding in further steps
* Many times we use to run our code repeatedly by taking different models as well
* Instead of checking every time with different models, we can create a function to select the best model
* It would be good to create one function to select the best model
* This function automatically test all models and finally returns the best model as per our requirement
* In this function, in very first step we are creating dictionary with XGBoost, Random Forest, Decision Tree, SVC, and a Multilayer Perceptron among others
* We can loop through these models, run the data through the pipeline for each model
* Use cross-validation to get good performance, reliability and validity of each model
* Store the model results in pandas DataFrame and print the output
* Finally selecting the BEST MODEL with highest ROC/AUC score

## Function explanation

* select_model(X, y, pipeline = None) is user defined function
* This function takes 2 non default parameters and 1 default parameter  

X (object) : Pandas dataframe containing X_train data

y (object) : Pandas dataframe containing y_train data

pipeline : Pipeline from model_pipeline()

In [38]:
def select_model(X, y, pipeline=None):

    classifiers = {}
    

    c_d4 = {"RandomForestClassifier": RandomForestClassifier()}
    classifiers.update(c_d4)

    c_d5 = {"DecisionTreeClassifier": DecisionTreeClassifier()}
    classifiers.update(c_d5)

    c_d9 = {"KNeighborsClassifier": KNeighborsClassifier()}
    classifiers.update(c_d9)

    c_d10 = {"RidgeClassifier": RidgeClassifier()}
    classifiers.update(c_d10)

    c_d13 = {"BernoulliNB": BernoulliNB()}
    classifiers.update(c_d13)

    c_d14 = {"SVC": SVC()}
    classifiers.update(c_d14)
   
    cols = ['model', 'run_time', 'roc_auc']
    df_models = pd.DataFrame(columns = cols)

    for key in classifiers:
        
        start_time = time.time()
        
        print()
        print("Step 12: model_pipeline run successfully on", key)

        pipeline = model_pipeline(X_train, classifiers[key])
        
        cv = cross_val_score(pipeline, X, y, cv=10, scoring='roc_auc')

        row = {'model': key,
               'run_time': format(round((time.time() - start_time)/60,2)),
               'roc_auc': cv.mean(),
        }

        df_models = pd.concat([df_models, pd.DataFrame([row])], ignore_index=True)
        
    df_models = df_models.sort_values(by='roc_auc', ascending=False)
	
    return df_models
    



In [41]:
# Once we call select_model function then we will get best model from all models

models = select_model(X_train, y_train)

models


Step 12: model_pipeline run successfully on RandomForestClassifier


C:\Users\Pratibha\AppData\Local\Temp\ipykernel_20048\544685488.py:43: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.
  df_models = pd.concat([df_models, pd.DataFrame([row])], ignore_index=True)



Step 12: model_pipeline run successfully on DecisionTreeClassifier

Step 12: model_pipeline run successfully on KNeighborsClassifier

Step 12: model_pipeline run successfully on RidgeClassifier

Step 12: model_pipeline run successfully on BernoulliNB

Step 12: model_pipeline run successfully on SVC


model run_time   roc_auc
5                     SVC      0.4  0.887121
0  RandomForestClassifier     0.89  0.885314
4             BernoulliNB     0.01  0.865736
3         RidgeClassifier     0.01  0.858738
2    KNeighborsClassifier     0.02  0.841365
1  DecisionTreeClassifier     0.04  0.734978

## The best model is
* The top performer was SVC(), which generated a ROC/AUC score of 0.88 on the training data, which is really good
* We’ll select this model as our best one and examine the results in a bit more detail to see how well it works

## Examine the performance of the best model

In [42]:
# Creating the model

selected_model = XGBClassifier()
bundled_pipeline = model_pipeline(X_train, selected_model)

# Training the model

bundled_pipeline.fit(X_train, y_train)

Pipeline(steps=[('preprocessor',
                 ColumnTransformer(remainder='passthrough',
                                   transformers=[('numeric',
                                                  SimpleImputer(strategy='constant'),
                                                  ['Administrative',
                                                   'Administrative_Duration',
                                                   'Informational',
                                                   'Informational_Duration',
                                                   'ProductRelated',
                                                   'ProductRelated_Duration',
                                                   'BounceRates', 'ExitRates',
                                                   'PageValues', 'SpecialDay',
                                                   'Month', 'OperatingSystems',
                                                   'Browser'...
                               feature_types=None, gamma=None, grow_policy=None,
                               importance_type=None,
                               interaction_constraints=None, learning_rate=None,
                               max_bin=None, max_cat_threshold=None,
                               max_cat_to_onehot=None, max_delta_step=None,
                               max_depth=None, max_leaves=None,
                               min_child_weight=None, missing=nan,
                               monotone_constraints=None, multi_strategy=None,
                               n_estimators=None, n_jobs=None,
                               num_parallel_tree=None, random_state=None, ...))])

## Prediction

In [43]:
y_pred = bundled_pipeline.predict(X_test)
print(y_pred)

[0 0 0 ... 0 0 0]


## ROC and AUC scores

In [45]:
roc_auc = roc_auc_score(y_test, y_pred)
accuracy = accuracy_score(y_test, y_pred)
f1_score = f1_score(y_test, y_pred)

print('ROC/AUC:', roc_auc)
print('Accuracy:', accuracy)
print('F1 score:', f1_score)

ROC/AUC: 0.7889905668575586
Accuracy: 0.8798111837327524
F1 score: 0.6323583857830434


## Classification report

In [46]:
classif_report = classification_report(y_test, y_pred)

classif_report

'              precision    recall  f1-score   support\n\n           0       0.93      0.92      0.93      6961\n           1       0.61      0.66      0.63      1301\n\n    accuracy                           0.88      8262\n   macro avg       0.77      0.79      0.78      8262\nweighted avg       0.88      0.88      0.88      8262\n'

## Final results and words

* xamining the confusion matrix for the selected model, it shows that we correctly predicted customers (93%) wouldn’t purchase during their session, and we correctly predicted that 63% of customers who would purchase during their sessions
* Clearly, there’s still more we could do, and the overall accuracy of 88%
* The results show that it’s possible to predict purchase intention from consumer behaviour data with a good degree of accuracy